controllers/util/k8sutil.go

```diff
@@ -289,3 +292,51 @@ func NewK8sClient(scheme *runtime.Scheme) (client.Client, error) {
 
 	return k8sClient, nil
 }
+
+func CreateOrReplace(ctx context.Context, c client.Client, obj client.Object, f controllerutil.MutateFn) error {
+	key := client.ObjectKeyFromObject(obj)
+	if err := c.Get(ctx, key, obj); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		if err := mutate(f, key, obj); err != nil {
+			return err
+		}
+		if err := c.Create(ctx, obj); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	existing := obj.DeepCopyObject()
+	if err := mutate(f, key, obj); err != nil {
+		return err
+	}
+
+	if equality.Semantic.DeepEqual(existing, obj) {
+		return nil
+	}
+
+	if err := c.Delete(ctx, obj); err != nil {
+		return err
+	}
+
+	// k8s doesn't allow us to create objects when resourceVersion is set, as we are DeepCopying the
+	// object, the resource version also gets copied, hence we need to set it to empty before creating it
+	obj.SetResourceVersion("")
+	if err := c.Create(ctx, obj); err != nil {
+		return err
+	}
+	return nil
+}
+
+// mutate wraps a MutateFn and applies validation to its result.
+func mutate(f controllerutil.MutateFn, key client.ObjectKey, obj client.Object) error {
+	if err := f(); err != nil {
+		return err
+	}
+	if newKey := client.ObjectKeyFromObject(obj); key != newKey {
```

Comment:
Should we not check this before calling a func?

Reply:
We have been using this in the client op from 4.18 and it's working as expected